### PR TITLE
Tracking events domain objects

### PR DESCRIPTION
**Pull request by dimamo5**
Depends on #3 

This PR adds tracking events objects. Now the tracking function uses these objects instead of just a `JSONObject`. To allow the user to send their own types I created a `CustomTrackingEvent` that accepts a JSONObject and only validates its `type`, `clientId` and `siteId`.

### CODE REVIEW

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/AddToCart.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class AddToCart(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/AddToCart.kt]: This should be an array of `Product`

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductClick.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ProductClick(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductClick.kt]: Again, array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductCustomization.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ProductCustomization(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null,
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductCustomization.kt]: Array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductCustomization.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @Optional val productCustomization: Customization? = null
+) : TrackingEvent("productCustomization", siteId, clientId) {
+    override fun serialize(): String =
+        jsonSerilizer().stringify(serializer(), this)
+
+    companion object Properties {
+        @Serializable
+        data class Customization(
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductCustomization.kt]: All the properties of `Customization` can be `@Optional`

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductView.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ProductView(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductView.kt]: Array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductViewDetails.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ProductViewDetails(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductViewDetails.kt]: Array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductFeedback.kt [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ProductFeedback(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null,
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/ProductFeedback.kt]: Array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/RemoveFromCart.kt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class RemoveFromCart(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/RemoveFromCart.kt]: Array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/Subscription.kt [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Subscription(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    @Optional val product: Product? = null,
```
Comment (duartepinto) [inline velocidi-sdk/src/main/kotlin/com/velocidi/events/Subscription.kt]: Array of `Product` 😅

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/Refund.kt [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package com.velocidi.events
+
+import kotlinx.serialization.Optional
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Refund(
+    @Optional override val siteId: String? = null,
+    @Optional override val clientId: String? = null,
+    val refundType: String,
```
[the remark this answers is not in the document]
Comment (dimamo5) [reply velocidi-sdk/src/main/kotlin/com/velocidi/events/Refund.kt]: I don't want to extend our inheritance in the domain models. Kotlin serialization has some problems with it.
I think it's ok to leave it like as a `String` for now. We should probably revisit this later

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/CustomTrackingEvent.kt [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package com.velocidi.events
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.StringDescriptor
+import org.json.JSONObject
+
+@Serializable
+class CustomTrackingEvent(@Transient val json: JSONObject = JSONObject()) :
```
[the remark this answers is not in the document]
Comment (dimamo5) [reply velocidi-sdk/src/main/kotlin/com/velocidi/events/CustomTrackingEvent.kt]: Check 1602290

File: README.md [hunk 12]
```diff
@@ -70,18 +70,25 @@ class SampleApplication : Application() {
...
-                  "siteId": "1",
-                  "clientId": "client1",
-                  "location": "MainActivity"
-                }
-            """.trimIndent()
-Velocidi.track(JSONObject(event))
+This method is expecting a tracking event with the event details. For more information check our [documentation](https://docs.velocidi.com/knowledgebase/web-and-e-commerce/)
+
+```
+Velocidi.getInstance().track(PageView("pageView", "MobileApp", "client1"))
```
[the remark this answers is not in the document]
Comment (dimamo5) [reply README.md]: Changed in 6305e05

**Comment by dimamo5** [comment]
> Apart from what I left as comments, I also think that siteId and clientId should not be optional when instantiating an instance of TrackingEvent. All subclasses should require those fields in the constructor.

Improved in 38af754

File: velocidi-sdk/src/main/kotlin/com/velocidi/events/Refund.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (duartepinto) [reply velocidi-sdk/src/main/kotlin/com/velocidi/events/Refund.kt]: What about the `refundType` being an enum?

[the remark this answers is not in the document]
Comment (dimamo5) [reply velocidi-sdk/src/main/kotlin/com/velocidi/events/Refund.kt]: It can be an enum. But I will have to make a custom serilizer

**Comment by duartepinto** [comment]
> Also, personal opinion, don't really like that we have all the possible properties declared in the class constructor. This leads to really big constructors. But this is more of a personal thing, don't have a really strong opinion about this.

About this, due to the offline conversation that we had, If in the future we drop the `siteId` and the `clientId`, in order to keep backward compatibility, we will still have to provide a constructor with those parameters, leading us to have a really big constructor just for backwards compatibility. Having such big constructor is not easy to manage in the long run. I would prefer, for now, to only have the constructor with `siteId`, `typeId` and other required properties for each model class so that it is easier to drop them or change them in the future, should the necessity arise.

**Comment by dimamo5** [comment]
I've been doing some testing and `kotlinx.serialization` doesn't like having properties defined in the class body. It would serialize them but to ones in the parent class.

For example: 
```
data class AddToCart(
    override val siteId: String,
    override val clientId: String
) : TrackingEvent("addToCart") {
    var products: List<Product> = emptyList()

    override fun serialize(): String =
        jsonSerilizer().stringify(serializer(), this)
}
```
In this case, it would only serialize siteId, clientId and `products` but not `type`.

I also tried changing `TrackingEvent` to an interface or making `type` an abstract val but neither of them worked.

I think our only solution is leaving it like it is.

**Comment by duartepinto** [comment]
> I've been doing some testing and `kotlinx.serialization` doesn't like having properties defined in the class body. It would serialize them but to ones in the parent class.

What about replacing `kotlinx.serialization` with something like [gson](https://github.com/google/gson) which is a bit more stable? Performance wise, `gson` is also quite fast and is a bit more mature than `kotlinx.serialization` (which still hasn't reached version 1.X)

**Comment by dimamo5** [comment]
I think with Gson we couldn't have inheritance and we could only generate JSON from `data class`.
We picked `kotlinx.serialization` because it's officially supported by Kotlin and is already shipped with the Kotlin compiler.

If you think it's worth I can explore `gson`.

**Comment by duartepinto** [comment]
> If you think it's worth I can explore `gson`.

Yes, please. Just give it a quick look and see if it is feasible.

**Comment by dimamo5** [comment]
After some R&D I found that Gson works really well with Kotlin. I decided to migrate in d9c1bcb and applied your suggestion regarding "big constructors" in 541184f. I also prefer the way custom serializers are built in Gson.

**Comment by duartepinto** [comment]
>  applied your suggestion regarding "big constructors" in [faedd38](https://github.com/velocidi/velocidi-android-sdk/commit/faedd380e34b067a833853780aa9663fd75b0bfa)

Sorry for nitpicking on this, but can you do the same for `Product` `ProductCustomization` and `Transaction`?

**Comment by dimamo5** [comment]
Improved in 9c89eb0

**Pull request merged**